autosklearn/data/validation.py

```diff
@@ -472,3 +472,58 @@ def decode_target(
             return self.target_encoder.inverse_transform(y)
         else:
             return self.target_encoder.named_transformers_['ordinalencoder'].inverse_transform(y)
+
+    def join_and_check(
+        self,
+        y: Union[np.ndarray, pd.DataFrame],
+        y_test: Union[np.ndarray, pd.DataFrame]
+    ) -> Union[np.ndarray, pd.DataFrame]:
+        """
+        This method checks for basic input quality before
+        merging the inputs to Auto-sklearn for a common
+        encoding if needed
+        """
+
+        # We expect same type of object
+        if type(y) != type(y_test):
+            raise ValueError(
+                "Train and test targets must be of the same type, yet y={} and y_test={}"
+                "".format(
+                    type(y),
+                    type(y_test)
+                )
+            )
+
+        if isinstance(y, pd.DataFrame):
+            # The have to have the same columns and types
+            if y.columns != y_test.columns:
+                raise ValueError(
+                    "Train and test targets must both have the same columns, yet "
+                    "y={} and y_test={} ".format(
+                        type(y),
+                        type(y_test)
+                    )
+                )
+
+            if list(y.dtypes) != list(y_test.dtypes):
+                raise ValueError("Train and test targets must both have the same dtypes")
+
+            return pd.concat([y, y_test], ignore_index=True, sort=False)
+        elif isinstance(y, np.ndarray):
+            # The have to have the same columns and types
+            if len(y.shape) != len(y_test.shape) \
+                    or (len(y.shape) > 1 and (y.shape[2] != y_test.shape[2])):
+                raise ValueError("Train and test targets must have the same dimensionality")
+
+            if y.dtype != y_test.dtype:
+                raise ValueError("Train and test targets must both have the same dtype")
+
+            return np.concatenate((y, y_test))
+        elif isinstance(y, list):
+            # Provide flexibility in the list. When transformed to np.ndarray
+            # further checks are performed downstream
+            return y + y_test
+        else:
+            # If reached this point, it is an object we don't support on
```

Comment:
Wouldn't an exception be more appropriate here if we don't support the object?

Reply:
It is not so much as we don't support the object, but it is for the specific scenario of a sparse array, which we don't encode and just let pass to sklearn checks (because of course you can't have sparse categorical scipy arrays, the closest is integer which by itself is encoded)

Currently, the message is:
# If reached this point, it is an object we don't support on
 # encoding, so we can let it pass. For instance a sparse int array

Does it make sense to just change it to the following or you consider further action is required?
 # If reached this point, it is an object we should not encode (sparse data)
 # In this scenario, creating a join target vector for encoding is not necessary

Reply:
(The other possibility is to just use hstack from scipy sparse, and error in any other case. This will give an extra check to make sure no unexpected  type is passed)

Reply:
This approach is fine. Later on the sparse data will be rejected anyway as we don't support sparse outputs.